### ISSUE TITLE
[string-remove-widows] how to ignore mso comments correctly?

### ISSUE THREAD

**Issue by cossssmin** [codsen/codsen#55]
**Package's name**
string-remove-widows

So there's basically two types of MSO comments, that could be simplfied to these:

```html
<!--[if mso]>
  <div>Show this in all Outlook versions</div>
<![endif]-->
```

and:

```html
<!--[if !mso]><!-->
  <div>All Outlooks will ignore this</div>
<!--<![endif]-->
```

I've set `minWordCount: 2` and added the following heads/tails in addition to the Jinja ones:

```js
{
  heads: '<!--[',
  tails: ']>'
},
{
  heads: '<![',
  tails: ']--><'
}
```

And it's _mostly_ working, however it's not enough because this:

```html
<!--[if mso]>
  <p>A paragraph inside an Outlook MSO&nbsp;comment</p>
  <p>unescaped {{ foo }}</p>
<![endif]-->
```

... ends up getting a `nbsp;` between the paragraph tags:

```html
<!--[if mso]>
  <p>A paragraph inside an Outlook MSO&nbsp;comment</p>&nbsp;<p>unescaped {{ foo }}</p>
<![endif]-->
```

Any advice on how to correctly ignore MSO comments? Thanks!

**Comment by revelt** [comment]
hi Cosmin!

The second object's tails look suspicious, `]--><` last character. Try universal heads and tails:

```
{
  heads: "<!--[",
  tails: "]-->",
}
```

I've just checked

```js
test(`19 - multi-paragraph combo with jinja`, () => {
  let source = `<!--[if mso]>
  <p>A paragraph inside an Outlook MSO&nbsp;comment</p>
  <p>unescaped {{ foo }}</p>
<![endif]-->`;
  let res = removeWidows(source, {
    minWordCount: 2,
    ignore: [
      {
        heads: "<!--[",
        tails: "]-->",
      },
      {
        heads: "{#",
        tails: "#}",
      },
      {
        heads: "{%",
        tails: "%}",
      },
      {
        heads: "{{",
        tails: "}}",
      },
    ],
  });
  equal(res.res, source, "19");
});


test(`20 - multiparagraph combo with jinja`, () => {
  let source = `<!--[if !mso]><!-->
  <p>A paragraph inside an Outlook MSO&nbsp;comment</p>
  <p>unescaped {{ foo }}</p>
<!--<![endif]-->`;
  let res = removeWidows(source, {
    minWordCount: 2,
    ignore: [
      {
        heads: "<!--[",
        tails: "]-->",
      },
      {
        heads: "{#",
        tails: "#}",
      },
      {
        heads: "{%",
        tails: "%}",
      },
      {
        heads: "{{",
        tails: "}}",
      },
    ],
  });
  equal(res.res, source, "20");
});
```

both pass

**Comment by cossssmin** [comment]
Sorry, I think I explained poorly.

So when you have `minWordCount: 2`, the library will transform this:

```html
<!--[if mso]>
  <div>Show this in all Outlook versions</div>
<![endif]-->
```

... to this:

```html
<!--[if&nbsp;mso]>
  <div>Show this in all Outlook&nbsp;versions</div>
<![endif]-->
```

Notice the `&nbsp;` in the `if mso` statement - it will no longer work for targeting Outlook.

It should ideally do this:

```html
<!--[if mso]>
  <div>Show this in all Outlook&nbsp;versions</div>
<![endif]-->
```

**Comment by revelt** [comment]
The content between heads and tails is ignored, so this div would not get touched:

```js
test(`21 - multiparagraph combo with jinja`, () => {
  let source = `<!--[if mso]>
  <div>Show this in all Outlook versions</div>
<![endif]-->`;
  let res = removeWidows(source, {
    minWordCount: 2,
    ignore: [
      {
        heads: "<!--[",
        tails: "]-->",
      },
      {
        heads: "{#",
        tails: "#}",
      },
      {
        heads: "{%",
        tails: "%}",
      },
      {
        heads: "{{",
        tails: "}}",
      },
    ],
  });
  equal(res.res, source, "21");
});
```

If we want to process these code blocks, the best solution is to parse, traverse the AST and process each text node. But then, the challenge is recognising all email templating languages (because unrecognised-ones will break the parser)... Personally, I solve this challenge by using two-level rendering; ie. Nunjucks is rendering Jinja and the content ("John Smith") is separate from the dynamic templating bits ("if - else" or "{{ data.firstName}}"). This way, I'm processing the JSON contents, rather than the final template.

Cosmin, did you try those universal heads and tails, the

```
{
  heads: "<!--[",
  tails: "]-->",
},
```

?

Basically, it seems the missing part is parser; this library is not ready to process HTML trees, it can only recognise styling tags such as `<strong>` etc, but not nested structures such as `<div>`. It would be unpractical to code this up; there are existing libraries that do it.

Do you see?
